### ISSUE TITLE
streamText: add callback for handling aborted streams

### ISSUE THREAD

**Issue by jvandenaardweg** [vercel/ai#309]
Now we have stream cancellation working properly (https://github.com/vercel-labs/ai/issues/90#issuecomment-1618915409 ) I think we need a way to detect and handle the cancellation server side.

In the current situation when a user cancels a stream, the [server errors](https://github.com/vercel-labs/ai/issues/90#issuecomment-1619424795) and there's no way to catch the cancellation and handle what should happen on the server. But I want to know how many tokens are generated when the user cancels the stream, so I can then store that information in my external database.

Currently we have `onStart`, `onCompletion` and `onToken` callback methods on `OpenAIStream()` / `AIStream()`. Ideally a new callback method would be great, something like `onStop` (to follow the `stop()` method naming in the React Hooks) or `onCancel`, which passes through the completion text at that point and allows to calculate the tokens and write that to a external database.

Or a different way to just catch the abort error in our (Edge) Functions.

From my understanding, I guess we need to postpone the abort until whatever method is doing something async is done, otherwise the serverless function might not exist anymore.

Is this even possible?

**Comment by hassanbazzi** [comment]
Having the same issue. This is very important.

**Comment by gartia** [comment]
This would be nice. to have an OnStop. I am finding it annoying to deal with function calls, when i want to abort after a function call. But i getting errors from using abort controllers via server methods.

**Comment by lassegit** [comment]
Is it really necessary to have all these callbacks? Having `onStop(completion, error)` call should be enough. My problem is that I don't have a callback when the abort signal is called. I need a callback to be able to track token usage.

**Comment by GorvGoyl** [comment]
What are some current workarounds till the team fixes the issue?

**Comment by GorvGoyl** [comment]
related issue https://github.com/vercel/next.js/discussions/48381

**Comment by peterdresslar** [comment]
Hello @GorvGoyl, are you calling the OpenAI using the OpenAI Node library (as documented in the examples from this SDK) or using your own custom logic to access the API?

**Comment by GorvGoyl** [comment]
> Hello @GorvGoyl, are you calling the OpenAI API using the OpenAI Node library (as documented in the examples from this SDK) or using your own custom logic to access the API?

I'm using vercel edge function like this:

```
 const completion = await openai.chat.completions.create({
            model: 'gpt-3.5-turbo',
            stream: true,
            messages: [{ role: 'user', content: question }],
        });

        const stream = OpenAIStream(completion, {...});
        const streamResponse = new StreamingTextResponse(stream);
            return streamResponse;
```

and then in the frontend (chrome extension) I'm using `ReadableStream` and `TextDecoder` to parse (can't use useChat or useCompletion in extension)

**Comment by jvandenaardweg** [comment]
> What are some current workarounds till the team fixes the issue?

I mentioned 2 work arounds here: https://github.com/vercel/ai/issues/90#issuecomment-1700960926

> 1. When a user initiates a cancel from your UI, send the AI generated text to an API that tracks the token/character/word usage of that text. Probable downside is when a user navigates away from your site (or closes the browser), then the API call will not be fired.
> 2. Track each generated token using `onToken` on the server and after a certain timeout just assume its cancelled. Also track the completion, so you can cancel them out. But for this you need to keep a record in a Redis store or something. Send each token to that store (that are a lot of requests depending on the size of text being generated). Because once the stream is done, the Edge Function is terminated and your timeout code will not run. Then I guess you need to poll Redis each minute or so and pull that usage info into your main database.
> 
> I currently have option 1 in place in my app. Not optimal, but in terms of dev time the easiest I think.

> related issue [vercel/next.js#48381](https://github.com/vercel/next.js/discussions/48381)

Yes, that's the same issue. Once aborted, you get a uncaught exception, which you can't catch.

**Comment by yvngjordi** [comment]
Still having the same problem. Guess Vercel SDK is just limited

**Comment by lgrammel** [comment]
The AI SDK functions support a `abortSignal` option that you can use to forward the abort.

**Comment by verrannt** [comment]
Could you elaborate a bit how to achieve the desired functionality using `abortSignal`? I am facing the same problem as OP and it is not at all clear to me how I would go about implementing it with `abortSignal`.

Or anyone else who has successfully implemented a workaround to achieve something like an `onStop` callback for manually stopped responses?

**Comment by xl0** [comment]
@lgrammel It Would be nice if abortSignal could trigger onFinish and pass as much data as possible to it. This would make it easier to save aborted messages to the database.

**Comment by lgrammel** [comment]
@xl0 this will be problematic because of typesafety and cause a lot of optionality checks. I would prefer something like`onAbort` - would that work for you?

**Comment by xl0** [comment]
@lgrammel onAbort would work perfectly fine for me.

**Comment by Thimows** [comment]
@lgrammel Any update on supporting 'onAbort'?

**Comment by lgrammel** [comment]
We are currently considering calling onFinish with a special finish reason (`abort`). Still TBD whether we go that direction or with `onAbort`. No timeline yet/

**Comment by Kemerd** [comment]
Also running into this.. even if I setup a system to use abort signal, how would we be able to pass it to the stream on something like an edge function, where a new instance is created each time?

**Comment by andrewshvv** [comment]
I'm just curious—since onStop / onAbort isn’t currently supported and it’s not a priority, how is token usage tracked when a request is cancelled? I feel like there’s probably an obvious solution I’m overlooking, which might explain why it hasn’t been a focus.

**Comment by Malakasd748** [comment]
I think this is a very important feature. Without a proper way to handle aborted streams, the client side `stop()` function doesn't make any sense. I don't understand why this isn't prioritized.

**Comment by slooowshutter** [comment]
Any update on this?

**Comment by istarkov** [comment]
@lgrammel but onAbort is not called in case of user has cancelled the stream. 
The only way I found is to create intermediate TransformStream, and call abortController.abort inside cancel event.

i.e.

```js
  const cancelInterceptorStream = stream
    .pipeThrough(
      new TransformStream<UIDataChunk, UIDataChunk>({
        cancel(reason) {
          abortController.abort("Cancel stream received");
        },
        transform(chunk, controller) {
          controller.enqueue(chunk);
        },
      }),
    );

  return createUIMessageStreamResponse({ stream: cancelInterceptorStream });
```

In no other way cancellation of the response stream is somehow intercepted

And seems like if response is cancelled during tool calling, abort is not executed at all. No events are produced at all.

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Bug**  
Confidence: **High**

### Rationale
Although the issue originally proposed a new cancellation callback, it describes concrete broken cancellation behavior, and later comments report that the available `onAbort` mechanism does not fire as expected.

### Evidence
- Cancelling a stream reportedly causes an uncaught server exception that cannot be handled.
- The latest comment reports that the existing `onAbort` callback is not invoked when the user cancels the response stream.
- Cancellation during tool calls reportedly produces no abort event, requiring a custom `TransformStream` workaround.
